### PR TITLE
multiple references in an analysis

### PR DESCRIPTION
**Pull request by antgonza**
This is the replacement for https://github.com/biocore/qiita/pull/1660

### CODE REVIEW

**Comment by antgonza** [comment]
Ready for review.

File: qiita_db/analysis.py [hunk 1]
```diff
@@ -808,81 +808,113 @@ def build_files(self,
...
 
-    def _build_biom_tables(self, samples, rarefaction_depth=None,
+    def _build_biom_tables(self, grouped_samples, rarefaction_depth=None,
                            rename_dup_samples=False):
         """Build tables and add them to the analysis"""
         with qdb.sql_connection.TRN:
             base_fp = qdb.util.get_work_base_dir()
 
-            # this assumes that there is only one reference/pipeline for each
-            # data_type issue #164
```
Comment (ElDeveloper) [inline qiita_db/analysis.py]: Since it seems like this issue will be closed, can you make a comment in an upcoming commit message (if there's one) to close this issue automatically?

Comment (antgonza) [inline qiita_db/analysis.py]: Yes.

File: qiita_db/analysis.py [hunk 2]
```diff
@@ -808,81 +808,113 @@ def build_files(self,
...
+                    # artifacts from multiple bioms and even then we might
+                    # only have one biom
+                    biom_table_fp = None
+                    for _, fp, fp_type in artifact.filepaths:
+                        if fp_type == 'biom':
+                            biom_table_fp = fp
+                            break
+                    if not biom_table_fp:
+                        raise RuntimeError(
+                            "Artifact %s do not have a biom table associated"
```
Comment (ElDeveloper) [inline qiita_db/analysis.py]: do -> does

File: qiita_db/analysis.py [hunk 3]
```diff
@@ -808,81 +808,113 @@ def build_files(self,
...
+                            % aid)
+
+                    # loading the found biom table
+                    biom_table = load_table(biom_table_fp)
+                    # filtering samples to keep those selected by the user
+                    biom_table_samples = set(biom_table.ids())
+                    selected_samples = biom_table_samples.intersection(samples)
+                    biom_table.filter(selected_samples, axis='sample',
```
Comment (ElDeveloper) [inline qiita_db/analysis.py]: Although very unlikely, this could return an empty table, which would still be a valid object, but would otherwise useless for everything processing related right? Do we want to do any check for that, maybe once we have the final "merged" table?

File: qiita_db/analysis.py [hunk 4]
```diff
@@ -808,81 +808,113 @@ def build_files(self,
...
-                    new_tables[data_type] = biom_table
-                else:
-                    new_tables[data_type] = \
-                        new_tables[data_type].merge(biom_table)
+                study_md = {'study': artifact.study.title,
+                            'artifact_ids': ', '.join(artifact_ids),
+                            'reference_id': reference_id,
+                            'command_id': command_id}
+                samples_md = {sid: study_md for sid in new_table.ids()}
+                new_table.add_metadata(samples_md, axis='sample')
```
Comment (ElDeveloper) [inline qiita_db/analysis.py]: :+1:

File: qiita_db/analysis.py (hunk 3 above)
Comment (antgonza) [inline qiita_db/analysis.py]: Excellent point, gonna add a count to the number of samples and add a test. Thanks!

**Comment by ElDeveloper** [comment]
:+1: looks good!

File: qiita_db/analysis.py [hunk 5]
```diff
@@ -808,81 +808,127 @@ def build_files(self,
...
+                    biom_table = load_table(biom_table_fp)
+                    # filtering samples to keep those selected by the user
+                    biom_table_samples = set(biom_table.ids())
+                    selected_samples = biom_table_samples.intersection(samples)
+                    biom_table.filter(selected_samples, axis='sample',
+                                      inplace=True)
+                    if len(biom_table.ids()) == 0:
+                        raise RuntimeError(
+                            "All samples filtered out from Artifact %s due "
+                            "to sample selected" % aid)
```
Comment (ElDeveloper) [inline qiita_db/analysis.py]: `sample selected` -> `selected samples`?

**Comment by ElDeveloper** [comment]
:+1: all good on my end!

File: qiita_db/analysis.py [hunk 6]
```diff
@@ -808,81 +808,122 @@ def build_files(self,
...
+            if not rename_dup_samples:
+                for k, v in viewitems(grouped_samples):
+                    # this element only has one table, continue
+                    if len(v) == 1:
+                        continue
+                    dup_samples = set()
+                    for k, v in v:
+                        v = set(v)
+                        if dup_samples & v:
+                            rename_dup_samples = (True and
```
Comment (josenavas) [inline qiita_db/analysis.py]: `rename_dup_samples = merge_duplicated_sample_ids`

File: qiita_db/analysis.py [hunk 7]
```diff
@@ -808,81 +808,122 @@ def build_files(self,
...
+            # 2. if rename_dup_samples is still False, make sure that we don't
+            #    need to rename samples by checking that there are not
+            #    duplicated samples per group
+            if not rename_dup_samples:
+                for k, v in viewitems(grouped_samples):
+                    # this element only has one table, continue
+                    if len(v) == 1:
+                        continue
+                    dup_samples = set()
+                    for k, v in v:
```
Comment (josenavas) [inline qiita_db/analysis.py]: Re-defining `v`, can we use better variable namaes? It's hard to follow the code as it is.

File: qiita_db/analysis.py [hunk 8]
```diff
@@ -808,81 +808,122 @@ def build_files(self,
...
+                fn = "%d_analysis_dt-%s_r-%s_c-%s.biom" % (
+                    self._id, data_type, reference_id, command_id)
+                biom_fp = join(base_fp, fn)
                 with biom_open(biom_fp, 'w') as f:
-                    biom_table.to_hdf5(f, "Analysis %s Datatype %s" %
-                                       (self._id, dt))
-                self._add_file("%d_analysis_%s.biom" % (self._id, dt),
-                               "biom", data_type=dt)
+                    new_table.to_hdf5(
+                        f, "Analysis %d Datatype %s Reference %s Command %s" %
```
Comment (josenavas) [inline qiita_db/analysis.py]: If I understand correctly, this information is stored as sample metadata. I think it is ok to leave here too if you want, but should we also add "QIITA" at the beginning since that parameter is called "Generated by"?

**Comment by josenavas** [comment]
A couple of minor comments
